### PR TITLE
Add Gauge Metrics Instrument

### LINKED ISSUE #92625: [Feature Request]: Add synchronous gauge instrument
***Edited by @tarekgh***

OpenTelemetry has introduced the [`Gauge`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#gauge) instrument, designed to record non-additive values when changes occur. For example, it can measure the background noise level, where summing the values from multiple rooms would be nonsensical.

## Proposal

```C#
namespace System.Diagnostics.Metrics;

public class Meter : IDisposable
{
	public Gauge<T> CreateGauge<T>(string name) where T : struct {}

	public Gauge<T> CreateGauge<T>(string name, string? unit = null, string? description = null, IEnumerable<KeyValuePair<string, object?>>? tags = null) where T : struct {}
}

public sealed class Gauge<T> : Instrument<T> where T : struct
{
	public void Record(T value) {}

	public void Record(T value, KeyValuePair<string, object?> tag) { }

	public void Record(T value, KeyValuePair<string, object?> tag1, KeyValuePair<string, object?> tag2) { }

	public void Record(T value, KeyValuePair<string, object?> tag1, KeyValuePair<string, object?> tag2, KeyValuePair<string, object?> tag3) { }

	public void Record(T value, params ReadOnlySpan<KeyValuePair<string, object?>> tags) { }

	public void Record(T value, params KeyValuePair<string, object?>[] tags) { }

	public void Record(T value, in TagList tagList) { }
	
	//
	// Instrument<T> properties
	//
	
	public Meter Meter { get; }
        public string Name { get; }
        public string? Description { get; }
        public string? Unit { get; }
        public IEnumerable<KeyValuePair<string, object?>>? Tags { get; }
        public bool Enabled { get; }
        public virtual bool IsObservable { get; }
}

```

## Usage Example

```C#
    Meter meter = new Meter("MeasurmentLibrary.Sound");
 
    Gauge<int> gauge= meter.CreateGauge<int>(name: "NoiseLevel", unit: "dB", description: "Background Noise Level"); // dB is Decibel, the sound intensity level unit
    
    gauge.Record(10, new TagList { "Room1", "dB"});
```

## Original Description

OpenTelemetry has now defined a synchronous gauge. If the OTel maintainers agree the spec is sufficiently stable we should implement it in System.Diagnostics.Metrics. Feel free to replace this stub with a more complete description of the new API surface.

In addition to adding the API, we also need to add support for it to the MetricsEventSource.

OTel issue: https://github.com/open-telemetry/opentelemetry-dotnet/issues/4805

cc @tarekgh

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes #92625

### CODE REVIEW

**Comment by tarekgh** [comment]
CC @noahfalk @reyang @cijothomas @CodeBlanch

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Gauge.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// The Gauge is an instrument used to record non-additive values whenever changes occur. For example, record each time the number of active threads changes.
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Gauge.cs:9]: I think the OTel spec uses temperature measurement as an example for Gauge and I think that would be better suited here.

The number of active threads is sometimes useful to handle additively - for example its reasonable to say "My fleet of computers uses a total of 745 threads to serve all my web requests." Temperature doesn't make sense to sum up - as in "My fleet of computers has CPU temperatures which sum to 3,821 degrees"

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Meter.cs [hunk 2]
```diff
@@ -143,6 +143,30 @@ private void Initialize(string name, string? version, IEnumerable<KeyValuePair<s
         public Counter<T> CreateCounter<T>(string name, string? unit, string? description, IEnumerable<KeyValuePair<string, object?>>? tags) where T : struct
                 => (Counter<T>)GetOrCreateInstrument<T>(typeof(Counter<T>), name, unit, description, tags, () => new Counter<T>(this, name, unit, description, tags));
 
+        /// <summary>
+        /// Create a metrics Gauge object.
+        /// </summary>
+        /// <param name="name">The instrument name. cannot be null.</param>
+        /// <remarks>
+        /// Gauge is an Instrument which used to record non-additive values.
+        /// Example uses for Gauge: record each time the number of active threads changes.
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Meter.cs:152]: Same example as above which I recommend we change.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Meter.cs [hunk 3]
```diff
@@ -143,6 +143,30 @@ private void Initialize(string name, string? version, IEnumerable<KeyValuePair<s
...
+        /// <summary>
+        /// Create a metrics Gauge object.
+        /// </summary>
+        /// <param name="name">The instrument name. cannot be null.</param>
+        /// <param name="unit">Optional instrument unit of measurements.</param>
+        /// <param name="description">Optional instrument description.</param>
+        /// <param name="tags">tags to attach to the counter.</param>
+        /// <remarks>
+        /// Gauge is an Instrument which used to record non-additive values.
+        /// Example uses for Gauge: record each time the number of active threads changes.
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Meter.cs:165]: Same example a last time.

**Review by noahfalk — APPROVED** [review]
LGTM other than a few nits in the doc comments. Thanks 👍

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Gauge.cs (hunk 1 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Gauge.cs]: Thanks, I used the example used in the OTel specs. `record the room background noise level value when changes occur.`. Thanks!

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#gauge

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Meter.cs [hunk 4]
```diff
@@ -143,6 +143,30 @@ private void Initialize(string name, string? version, IEnumerable<KeyValuePair<s
         public Counter<T> CreateCounter<T>(string name, string? unit, string? description, IEnumerable<KeyValuePair<string, object?>>? tags) where T : struct
                 => (Counter<T>)GetOrCreateInstrument<T>(typeof(Counter<T>), name, unit, description, tags, () => new Counter<T>(this, name, unit, description, tags));
 
+        /// <summary>
+        /// Create a metrics Gauge object.
```
Comment (cijothomas) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Meter.cs:147]:
```suggestion
        /// Creates a Gauge instrument, which can be used to record non-additive values.
```

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Meter.cs [hunk 5]
```diff
@@ -143,6 +143,30 @@ private void Initialize(string name, string? version, IEnumerable<KeyValuePair<s
...
+        /// </remarks>
+        public Gauge<T> CreateGauge<T>(string name) where T : struct => CreateGauge<T>(name, unit: null, description: null, tags: null);
+
+        /// <summary>
+        /// Create a metrics Gauge object.
+        /// </summary>
+        /// <param name="name">The instrument name. cannot be null.</param>
+        /// <param name="unit">Optional instrument unit of measurements.</param>
+        /// <param name="description">Optional instrument description.</param>
+        /// <param name="tags">tags to attach to the counter.</param>
```
Comment (cijothomas) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Meter.cs:162]:
```suggestion
        /// <param name="tags">tags to attach to the Gauge.</param>
```

**Pull request merged**